File: crates/python/mirascript/vm/lib/vm_global/debug.py
import sys
from ...operations import ToString_
from ...error import VmError


DEBUG = "\033[44;37m"
PANIC = "\033[41;37m"
RESET = "\033[0m"

DEBUG_PREFIX = f"{DEBUG} MiraScript {RESET}" if sys.stdout.isatty() else "[MiraScript]"
PANIC_PREFIX = f"{PANIC} MiraScript {RESET}" if sys.stdout.isatty() else "[MiraScript]"


def debug_print(*args):
    print(DEBUG_PREFIX, *args)


def panic(*msg):
    if not msg:
        print(PANIC_PREFIX)
    else:
        print(PANIC_PREFIX, *msg)

    error = f"panic: {ToString_(msg)}" if len(msg) > 0 else "panic"

    raise VmError("MiraScript panic", error)

File: crates/vm/src/standard_library/global/debug.rs
use crate::standard_library::global_builtin;
use crate::{MiraError, MiraValue, Result, Runtime, RuntimeErrorKind, operations};

fn debug_print(call: &mut Runtime, args: &[MiraValue]) -> Result<MiraValue> {
    let message = args
        .iter()
        .map(|value| operations::display(call, *value))
        .collect::<Vec<_>>()
        .join(" ");
    call.options().providers.debug(&message);
    Ok(MiraValue::NIL)
}

fn panic(call: &mut Runtime, args: &[MiraValue]) -> Result<MiraValue> {
    let message = args
        .first()
        .map(|value| operations::to_string(call, *value))
        .transpose()?
        .unwrap_or_else(|| "MiraScript panic".into());
    Err(MiraError::runtime(RuntimeErrorKind::UserMessage {
        message,
    }))
}

pub(super) fn install(context: &mut Runtime) {
    global_builtin!(
        context,
        fn debug_print: debug_print;
        fn panic: panic;
    );
}
